Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp

```diff
@@ -0,0 +1,484 @@
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebAutomationSession.h"
+
+#include "WebAutomationSessionLibWPE.h"
+#include "WebAutomationSessionMacros.h"
+#include "WebEventModifier.h"
+#include "WebPageProxy.h"
+#include <wpe/wpe.h>
+#include <wtf/glib/GUniquePtr.h>
+
+
+#if ENABLE(WPE_PLATFORM)
+#include "GRefPtrWPE.h"
+#include <wpe/wpe-platform.h>
+#endif
+
+
+
+namespace WebKit {
+
+// Called by platform-indendent code to convert the current platform-dependent raw modifiers into generic WebEventModifiers,
+// which will be passed to the platformSimulateFooBarInteraction methods.
+OptionSet<WebEventModifier> WebAutomationSession::platformWebModifiersFromRaw(WebPageProxy& page, unsigned modifiers)
+{
+    OptionSet<WebEventModifier> webModifiers;
+
+#if ENABLE(WPE_PLATFORM)
+    if (page.wpeView()) {
+        if (modifiers & WPE_MODIFIER_KEYBOARD_ALT)
+            webModifiers.add(WebEventModifier::AltKey);
+        if (modifiers & WPE_MODIFIER_KEYBOARD_META)
+            webModifiers.add(WebEventModifier::MetaKey);
+        if (modifiers & WPE_MODIFIER_KEYBOARD_CONTROL)
+            webModifiers.add(WebEventModifier::ControlKey);
+        if (modifiers & WPE_MODIFIER_KEYBOARD_SHIFT)
+            webModifiers.add(WebEventModifier::ShiftKey);
+        if (modifiers & WPE_MODIFIER_KEYBOARD_CAPS_LOCK)
+            webModifiers.add(WebEventModifier::CapsLockKey);
+    } else {
+#endif
+        if (modifiers & wpe_input_keyboard_modifier_alt)
+            webModifiers.add(WebEventModifier::AltKey);
+        if (modifiers & wpe_input_keyboard_modifier_meta)
+            webModifiers.add(WebEventModifier::MetaKey);
+        if (modifiers & wpe_input_keyboard_modifier_control)
+            webModifiers.add(WebEventModifier::ControlKey);
+        if (modifiers & wpe_input_keyboard_modifier_shift)
+            webModifiers.add(WebEventModifier::ShiftKey);
+        // libWPE has no Caps Lock modifier.
+#if ENABLE(WPE_PLATFORM)
+    }
+#endif
+
+    return webModifiers;
+}
+
+#if ENABLE(WPE_PLATFORM)
+
+static uint32_t modifiersToEventState(OptionSet<WebEventModifier> modifiers)
+{
+    uint32_t state = 0;
+    if (modifiers.contains(WebEventModifier::ControlKey))
+        state |= WPE_MODIFIER_KEYBOARD_CONTROL;
+    if (modifiers.contains(WebEventModifier::ShiftKey))
+        state |= WPE_MODIFIER_KEYBOARD_SHIFT;
+    if (modifiers.contains(WebEventModifier::AltKey))
+        state |= WPE_MODIFIER_KEYBOARD_ALT;
```

Comment:
New api supports caps lock.

Reply:
Added capslock and meta modifiers for new api